### PR TITLE
Mirror changes from mono/coreclr 

### PR DESCRIPTION
**Pull request by filipnavara**
Fix up of #17506

### CODE REVIEW

File: mono/mini/intrinsics.c [hunk 1]
```diff
@@ -575,6 +575,11 @@ emit_unsafe_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignatu
 		MONO_ADD_INS (cfg->cbb, ins);
 		return ins;
 	}
+	else if (!strcmp (cmethod->name, "SkipInit")) {
+ 		MONO_INST_NEW (cfg, ins, OP_NOP);
```
Comment (EgorBo) [inline mono/mini/intrinsics.c]: Is OP_NOP needed here actually?

**Comment by vargaz** [comment]
Note that mono will still zero-init these locals.

File: mono/mini/intrinsics.c (hunk 1 above)
Comment (vargaz) [reply mono/mini/intrinsics.c]: The function needs to return something, its going to be deleted later.

**Comment by filipnavara** [comment]
It's a performance optimization so not a big deal.

**Comment by filipnavara** [comment]
@monojenkins build OS X x64 C++

**Comment by tannergooding** [comment]
CoreCLR will still zero-init these as well. This only impacts C# definite assignment.

The JIT zeroing is controlled via the `.locals init` flag and will be controllable via the `SkipLocalsInitAttribute`.

**Comment by filipnavara** [comment]
@tannergooding I believe CoreLib in CoreCLR uses ILLinker (or some other tool) to strip `.locals init`. Mono does not do that, in generally also does the initialization despite the lack of `.locals init` to make debugging easier.

**Comment by marek-safar** [comment]
@filipnavara Mono could do strip the .locals init but we are not yet aligned on the build setups to make it easy to do

**Comment by tannergooding** [comment]
> I believe CoreLib in CoreCLR uses ILLinker

Yes. But there is also a compiler feature coming in for C# to directly support a new attribute `[SkipLocalsInit]` which does the same thing. Corelib will likely switch over to the language feature once that is in (as it won't require IL rewriting).

So:
* `Unsafe.SkipInit` allows bypassing language definite assignment
* `SkipLocalsInit` will allow bypassing runtime zeroing of locals (but not GC references, which must be zeroed anyways)

The latter removes double writes to most locals, since C# requires them to be definitely assigned before first use anyways (the exception being `stackalloc`). The former allows you to bypass definite assignment rules for scenarios like `C-style union types`, where the struct has multiple fields overlapping the same underlying memory location (via `FieldOffset`).

> In general, it also does the initialization despite the lack of .locals init to make debugging easier.

Even in Release mode? That seems non-ideal (especially for AOT scenarios).

**Comment by jaykrell** [comment]
I don't think there is Release and Debug mode here. Is there?
I just changed it btw so the IL Link removal of flag, or newer forms, should be effective.

https://github.com/mono/mono/commit/f0ff9f3860b50f3ed857b9b820d5adce5cbd7f10#diff-e7526071878d708680e4a8c5fa5a46ed

**Pull request merged**